api/api.go

```diff
@@ -48,44 +47,67 @@ func (a *API) ListenAndServe(hostAndPort string) {
 }
 
 // NewAPI instantiates a new REST API
-func NewAPI(config *conf.Configuration, db storage.Connection, mailer mailer.Mailer) *API {
-	return NewAPIWithVersion(context.Background(), config, db, mailer, defaultVersion)
+func NewAPI(globalConfig *conf.GlobalConfiguration, db storage.Connection) *API {
+	return NewAPIWithVersion(context.Background(), globalConfig, db, defaultVersion)
 }
 
 // NewAPIWithVersion creates a new REST API using the specified version
-func NewAPIWithVersion(ctx context.Context, config *conf.Configuration, db storage.Connection, mailer mailer.Mailer, version string) *API {
-	api := &API{config: config, db: db, mailer: mailer, version: version}
+func NewAPIWithVersion(ctx context.Context, globalConfig *conf.GlobalConfiguration, db storage.Connection, version string) *API {
+	api := &API{config: globalConfig, db: db, version: version}
 
 	r := newRouter()
 	r.Use(addRequestID)
 	r.UseBypass(newStructuredLogger(logrus.StandardLogger()))
 	r.Use(recoverer)
 
-	r.Get("/", api.Index)
-	r.Post("/signup", api.Signup)
-	r.Post("/recover", api.Recover)
-	r.Post("/verify", api.Verify)
-	r.Post("/token", api.Token)
-	r.With(api.requireAuthentication).Post("/logout", api.Logout)
-
-	r.Route("/user", func(r *router) {
-		r.Use(api.requireAuthentication)
-		r.Get("/", api.UserGet)
-		r.Put("/", api.UserUpdate)
-	})
-
-	r.Route("/admin", func(r *router) {
-		r.Use(addGetBody)
-		r.Use(api.requireAuthentication)
-		r.Use(api.requireAdmin)
-		r.Get("/users", api.adminUsers)
+	r.Route("/", func(r *router) {
+		if globalConfig.MultiInstanceMode {
+			r.Use(api.loadInstanceConfig)
+		}
 
-		r.Post("/user", api.adminUserCreate)
-		r.Get("/user", api.adminUserGet)
-		r.Put("/user", api.adminUserUpdate)
-		r.Delete("/user", api.adminUserDelete)
+		r.Post("/signup", api.Signup)
+		r.Post("/recover", api.Recover)
+		r.Post("/verify", api.Verify)
+		r.Post("/token", api.Token)
+		r.With(api.requireAuthentication).Post("/logout", api.Logout)
+
+		r.Route("/user", func(r *router) {
+			r.Use(api.requireAuthentication)
+			r.Get("/", api.UserGet)
+			r.Put("/", api.UserUpdate)
+		})
+
+		r.Route("/admin", func(r *router) {
+			r.Use(addGetBody)
+			r.Use(api.requireAuthentication)
+			r.Use(api.requireAdmin)
+
+			r.Get("/users", api.adminUsers)
+
+			r.Post("/user", api.adminUserCreate)
+			r.Get("/user", api.adminUserGet)
+			r.Put("/user", api.adminUserUpdate)
+			r.Delete("/user", api.adminUserDelete)
+		})
 	})
 
+	if globalConfig.MultiInstanceMode {
```

Comment:
At some point we have to know if it is multi-instance mode or not to ensure we have all the configuration we need (either from file or loaded for each request). This config value is not user-controllable, but set via the different entrypoint commands (`serve` vs `multi`).

Having the management API be on a separate port is more complicated (extra code, more complex deployment, etc.). My preference is to keep as-is unless there is a technical reason. I think some of the folks who implement their own microservices are going to have single port deployments too.